File: packages/plugin-sdk-python/src/bookclerk_plugin_sdk/guest_sql.py
"""Guest SQL classification — must match host ``guest_statement_kind``."""

from __future__ import annotations

import re
from typing import Literal

DbStatementKind = Literal["query", "execute", "select", "returning"]


def _ident_start(c: str) -> bool:
    return c.isalpha() or c == "_"


def _ident_cont(c: str) -> bool:
    return _ident_start(c) or c.isdigit()


def _skip_ws_comments(sql: str, start: int) -> int:
    i = start
    while i < len(sql):
        while i < len(sql) and sql[i].isspace():
            i += 1
        if i + 1 < len(sql) and sql[i : i + 2] == "--":
            i += 2
            while i < len(sql) and sql[i] != "\n":
                i += 1
            continue
        if i + 1 < len(sql) and sql[i : i + 2] == "/*":
            i += 2
            while i + 1 < len(sql) and sql[i : i + 1] != "*/":
                i += 1
            i = min(i + 2, len(sql))
            continue
        break
    return i


def _keyword_at(sql: str, i: int, kw: str) -> bool:
    if i + len(kw) > len(sql):
        return False
    if sql[i : i + len(kw)].lower() != kw.lower():
        return False
    before_ok = i == 0 or not _ident_cont(sql[i - 1])
    after = sql[i + len(kw)] if i + len(kw) < len(sql) else " "
    return before_ok and not _ident_cont(after)


def _skip_ident_or_quoted(sql: str, i: int) -> int | None:
    if i >= len(sql):
        return None
    c = sql[i]
    if c in ('"', "`", "["):
        end = "]" if c == "[" else c
        j = i + 1
        while j < len(sql):
            if sql[j] == end:
                if end != "]" and j + 1 < len(sql) and sql[j + 1] == end:
                    j += 2
                    continue
                return j + 1
            j += 1
        return None
    if _ident_start(c):
        j = i + 1
        while j < len(sql) and _ident_cont(sql[j]):
            j += 1
        return j
    return None


def _skip_balanced_parens(sql: str, start: int) -> int | None:
    if start >= len(sql) or sql[start] != "(":
        return None
    depth = 0
    end: int | None = None

    def step(slice_sql: str, idx: int) -> int:
        nonlocal depth, end
        if end is not None:
            return 1
        c = slice_sql[idx]
        if c == "(":
            depth += 1
        elif c == ")":
            depth -= 1
            if depth == 0:
                end = start + idx + 1
        return 1

    _for_each_unquoted(sql[start:], step)
    return end


def _sql_after_leading_ctes(sql: str) -> str:
    i = _skip_ws_comments(sql, 0)
    if not _keyword_at(sql, i, "WITH"):
        return sql
    i += 4
    i = _skip_ws_comments(sql, i)
    if _keyword_at(sql, i, "RECURSIVE"):
        i += 9
        i = _skip_ws_comments(sql, i)
    while True:
        nxt = _skip_ident_or_quoted(sql, i)
        if nxt is None:
            return sql
        i = _skip_ws_comments(sql, nxt)
        if i < len(sql) and sql[i] == "(":
            after = _skip_balanced_parens(sql, i)
            if after is None:
                return sql
            i = _skip_ws_comments(sql, after)
        if not _keyword_at(sql, i, "AS"):
            return sql
        i = _skip_ws_comments(sql, i + 2)
        if _keyword_at(sql, i, "NOT"):
            after_not = _skip_ws_comments(sql, i + 3)
            if _keyword_at(sql, after_not, "MATERIALIZED"):
                i = _skip_ws_comments(sql, after_not + 12)
        elif _keyword_at(sql, i, "MATERIALIZED"):
            i = _skip_ws_comments(sql, i + 12)
        if i >= len(sql) or sql[i] != "(":
            return sql
        after_body = _skip_balanced_parens(sql, i)
        if after_body is None:
            return sql
        i = _skip_ws_comments(sql, after_body)
        if i < len(sql) and sql[i] == ",":
            i = _skip_ws_comments(sql, i + 1)
            continue
        return sql[i:]


def _for_each_unquoted(sql: str, step) -> None:
    i = 0
    in_s = in_d = in_line = in_block = False
    while i < len(sql):
        c = sql[i]
        if in_line:
            if c == "\n":
                in_line = False
            i += 1
            continue
        if in_block:
            if c == "*" and i + 1 < len(sql) and sql[i + 1] == "/":
                in_block = False
                i += 2
                continue
            i += 1
            continue
        if in_s:
            if c == "'":
                if i + 1 < len(sql) and sql[i + 1] == "'":
                    i += 2
                    continue
                in_s = False
            i += 1
            continue
        if in_d:
            if c == '"':
                if i + 1 < len(sql) and sql[i + 1] == '"':
                    i += 2
                    continue
                in_d = False
            i += 1
            continue
        if i + 1 < len(sql) and sql[i : i + 2] == "--":
            in_line = True
            i += 2
            continue
        if i + 1 < len(sql) and sql[i : i + 2] == "/*":
            in_block = True
            i += 2
            continue
        if c == "'":
            in_s = True
            i += 1
            continue
        if c == '"':
            in_d = True
            i += 1
            continue
        n = max(1, step(sql, i))
        i += n


def _for_each_top_level_keyword(sql: str, on_keyword) -> None:
    depth = 0

    def step(slice_sql: str, idx: int) -> int:
        nonlocal depth
        c = slice_sql[idx]
        if c == "(":
            depth += 1
            return 1
        if c == ")":
            depth = max(0, depth - 1)
            return 1
        if depth == 0 and _ident_start(c):
            j = idx + 1
            while j < len(slice_sql) and _ident_cont(slice_sql[j]):
                j += 1
            on_keyword(idx, slice_sql[idx:j])
            return j - idx
        return 1

    _for_each_unquoted(sql, step)


def _has_top_level_keyword(sql: str, keyword: str) -> bool:
    want = keyword.upper()
    found = False

    def on_keyword(_idx: int, kw: str) -> None:
        nonlocal found
        if kw.upper() == want:
            found = True

    _for_each_top_level_keyword(sql, on_keyword)
    return found


def _first_top_level_keyword(sql: str) -> str | None:
    first: str | None = None

    def on_keyword(_idx: int, kw: str) -> None:
        nonlocal first
        if first is None:
            first = kw.upper()

    _for_each_top_level_keyword(sql, on_keyword)
    return first


def guest_statement_kind(sql: str) -> DbStatementKind:
    """Classify guest SQL kind to match the host canonical request hash."""
    main = _sql_after_leading_ctes(sql)
    if _has_top_level_keyword(main, "RETURNING"):
        return "returning"
    verb = _first_top_level_keyword(main)
    if verb in {"SELECT", "VALUES"}:
        return "select"
    return "execute"


def split_exec_queries(query: str) -> list[str]:
    """Split Cloudflare-style ``exec()`` newline-separated queries."""
    out: list[str] = []
    for line in query.split("\n"):
        trimmed = line.strip()
        if not trimmed:
            continue
        trimmed = re.sub(r";+\s*$", "", trimmed).strip()
        if trimmed:
            out.append(trimmed)
    return out

File: packages/plugin-sdk/src/guest-sql.ts
/**
 * Guest SQL statement classification — must match the host
 * `guest_statement_kind` in `bookclerk-plugin-abi`.
 */

import type { DbStatementKind } from "./db-execute.js";

function identStart(c: string): boolean {
  const code = c.charCodeAt(0);
  return (code >= 65 && code <= 90) || (code >= 97 && code <= 122) || c === "_";
}

function identCont(c: string): boolean {
  return identStart(c) || (c >= "0" && c <= "9");
}

function skipWsComments(sql: string, start: number): number {
  let i = start;
  while (i < sql.length) {
    while (i < sql.length && /\s/.test(sql[i]!)) {
      i += 1;
    }
    if (sql[i] === "-" && sql[i + 1] === "-") {
      i += 2;
      while (i < sql.length && sql[i] !== "\n") {
        i += 1;
      }
      continue;
    }
    if (sql[i] === "/" && sql[i + 1] === "*") {
      i += 2;
      while (i + 1 < sql.length && !(sql[i] === "*" && sql[i + 1] === "/")) {
        i += 1;
      }
      i = Math.min(i + 2, sql.length);
      continue;
    }
    break;
  }
  return i;
}

function keywordAt(sql: string, i: number, kw: string): boolean {
  if (i + kw.length > sql.length) {
    return false;
  }
  if (sql.slice(i, i + kw.length).toLowerCase() !== kw.toLowerCase()) {
    return false;
  }
  const beforeOk = i === 0 || !identCont(sql[i - 1]!);
  const after = sql[i + kw.length] ?? " ";
  return beforeOk && !identCont(after);
}

function skipIdentOrQuoted(sql: string, i: number): number | null {
  const c = sql[i];
  if (c === undefined) {
    return null;
  }
  if (c === '"' || c === "`" || c === "[") {
    const end = c === "[" ? "]" : c;
    let j = i + 1;
    while (j < sql.length) {
      if (sql[j] === end) {
        if (end !== "]" && sql[j + 1] === end) {
          j += 2;
          continue;
        }
        return j + 1;
      }
      j += 1;
    }
    return null;
  }
  if (identStart(c)) {
    let j = i + 1;
    while (j < sql.length && identCont(sql[j]!)) {
      j += 1;
    }
    return j;
  }
  return null;
}

function skipBalancedParens(sql: string, start: number): number | null {
  if (sql[start] !== "(") {
    return null;
  }
  let depth = 0;
  let end: number | null = null;
  forEachUnquoted(sql.slice(start), (slice, idx) => {
    if (end !== null) {
      return 1;
    }
    const c = slice[idx]!;
    if (c === "(") {
      depth += 1;
    } else if (c === ")") {
      depth -= 1;
      if (depth === 0) {
        end = start + idx + 1;
      }
    }
    return 1;
  });
  return end;
}

function sqlAfterLeadingCtes(sql: string): string {
  let i = skipWsComments(sql, 0);
  if (!keywordAt(sql, i, "WITH")) {
    return sql;
  }
  i += 4;
  i = skipWsComments(sql, i);
  if (keywordAt(sql, i, "RECURSIVE")) {
    i += 9;
    i = skipWsComments(sql, i);
  }
  for (;;) {
    const next = skipIdentOrQuoted(sql, i);
    if (next === null) {
      return sql;
    }
    i = skipWsComments(sql, next);
    if (sql[i] === "(") {
      const after = skipBalancedParens(sql, i);
      if (after === null) {
        return sql;
      }
      i = skipWsComments(sql, after);
    }
    if (!keywordAt(sql, i, "AS")) {
      return sql;
    }
    i = skipWsComments(sql, i + 2);
    if (keywordAt(sql, i, "NOT")) {
      const afterNot = skipWsComments(sql, i + 3);
      if (keywordAt(sql, afterNot, "MATERIALIZED")) {
        i = skipWsComments(sql, afterNot + 12);
      }
    } else if (keywordAt(sql, i, "MATERIALIZED")) {
      i = skipWsComments(sql, i + 12);
    }
    if (sql[i] !== "(") {
      return sql;
    }
    const afterBody = skipBalancedParens(sql, i);
    if (afterBody === null) {
      return sql;
    }
    i = skipWsComments(sql, afterBody);
    if (sql[i] === ",") {
      i = skipWsComments(sql, i + 1);
      continue;
    }
    return sql.slice(i);
  }
}

function forEachUnquoted(sql: string, step: (slice: string, index: number) => number): void {
  let i = 0;
  let inS = false;
  let inD = false;
  let inLine = false;
  let inBlock = false;
  while (i < sql.length) {
    const c = sql[i]!;
    if (inLine) {
      if (c === "\n") {
        inLine = false;
      }
      i += 1;
      continue;
    }
    if (inBlock) {
      if (c === "*" && sql[i + 1] === "/") {
        inBlock = false;
        i += 2;
        continue;
      }
      i += 1;
      continue;
    }
    if (inS) {
      if (c === "'") {
        if (sql[i + 1] === "'") {
          i += 2;
          continue;
        }
        inS = false;
      }
      i += 1;
      continue;
    }
    if (inD) {
      if (c === '"') {
        if (sql[i + 1] === '"') {
          i += 2;
          continue;
        }
        inD = false;
      }
      i += 1;
      continue;
    }
    if (c === "-" && sql[i + 1] === "-") {
      inLine = true;
      i += 2;
      continue;
    }
    if (c === "/" && sql[i + 1] === "*") {
      inBlock = true;
      i += 2;
      continue;
    }
    if (c === "'") {
      inS = true;
      i += 1;
      continue;
    }
    if (c === '"') {
      inD = true;
      i += 1;
      continue;
    }
    const n = Math.max(1, step(sql, i));
    i += n;
  }
}

function forEachTopLevelKeyword(sql: string, onKeyword: (index: number, kw: string) => void): void {
  let depth = 0;
  forEachUnquoted(sql, (slice, idx) => {
    const c = slice[idx]!;
    if (c === "(") {
      depth += 1;
      return 1;
    }
    if (c === ")") {
      depth = Math.max(0, depth - 1);
      return 1;
    }
    if (depth === 0 && identStart(c)) {
      let j = idx + 1;
      while (j < slice.length && identCont(slice[j]!)) {
        j += 1;
      }
      onKeyword(idx, slice.slice(idx, j));
      return j - idx;
    }
    return 1;
  });
}

function hasTopLevelKeyword(sql: string, keyword: string): boolean {
  const want = keyword.toUpperCase();
  let found = false;
  forEachTopLevelKeyword(sql, (_, kw) => {
    if (kw.toUpperCase() === want) {
      found = true;
    }
  });
  return found;
}

function firstTopLevelKeyword(sql: string): string | undefined {
  let first: string | undefined;
  forEachTopLevelKeyword(sql, (_, kw) => {
    if (first === undefined) {
      first = kw.toUpperCase();
    }
  });
  return first;
}

/**
 * Classify guest SQL kind to match the host canonical request hash.
 *
 * @param sql Canonical Bookclerk SQL (`?` placeholders).
 * @returns Cap'n `DbStatementKind` (`execute` | `select` | `returning`).
 */
export function guestStatementKind(sql: string): DbStatementKind {
  const main = sqlAfterLeadingCtes(sql);
  if (hasTopLevelKeyword(main, "RETURNING")) {
    return "returning";
  }
  const verb = firstTopLevelKeyword(main);
  if (verb === "SELECT" || verb === "VALUES") {
    return "select";
  }
  return "execute";
}

/**
 * Split Cloudflare-style `exec()` newline-separated queries.
 *
 * @param query Multi-statement SQL with one statement per line.
 * @returns Trimmed statements with trailing semicolons removed.
 */
export function splitExecQueries(query: string): string[] {
  return query
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
    .map((line) => stripTrailingSemicolons(line).trim())
    .filter((line) => line.length > 0);
}

/**
 * Remove trailing `;` characters with a linear scan (no backtracking regex).
 *
 * @param line Already-trimmed SQL line.
 * @returns Line without its trailing semicolon run.
 */
function stripTrailingSemicolons(line: string): string {
  let end = line.length;
  while (end > 0 && line[end - 1] === ";") {
    end -= 1;
  }
  return line.slice(0, end);
}
